### PR TITLE
NNFrames: Cache preprocessed Sample

### PR DESCRIPTION
**Pull request by hhbyyh**
Change of behavior: 
1. By default, cache the Samples after preprocessing. For most of the use cases, this should bring improvement on both speed and memory.
2. add option to allow users to cache the raw data only.

fix https://github.com/intel-analytics/zoo/issues/263

### CODE REVIEW

**Comment by hhbyyh** [comment]
cc @jason-dai @dding3 @jenniew

**Comment by hhbyyh** [comment]
also cc @Oliverccccct This should resolve your memory issue.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNEstimator.scala [hunk 1]
```diff
@@ -102,6 +102,15 @@ private[nnframes] trait TrainingParams[@specialized(Float, Double) T] extends Pa
    */
   final val l2GradientClippingParams = new FloatParam(this,
     "threshold for l2 norm gradient clipping", "l2GradientClippingParams")
+
+  /**
+   * whether to cache the Sample after preprocessing.
+   * Default: true
+   */
+  final val isSampleCached = new BooleanParam(
+    this, "isSampleCached", "isSampleCached")
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNEstimator.scala]: change to "cachingSample"?

Comment (hhbyyh) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNEstimator.scala]: Ack.

Comment (hhbyyh) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/nnframes/NNEstimator.scala]: getCachingSample() can be a little confusing...

**Comment by hhbyyh** [comment]
Only the new fine tune example need to be updated. Transfer Learning and inference example are not applicable.

**Comment by dding3** [comment]
LGTM.

**Comment by hhbyyh** [comment]
http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/486 passed.
Thanks for review. Merging to master.

**Pull request merged**